### PR TITLE
Handle line separators in CodeMirror

### PR DESCRIPTION
**Pull request by blink1073**
Addresses part of https://github.com/jupyterlab/jupyterlab/issues/2951.  Without this change, the CodeMirror buffer positions are incorrect, resulting in a mismatch between the buffer and our document model. 

We normalize `\r\n` and `\r` line endings to `\n`, and mark the file as dirty if we change the line endings.

Note that upon pasting text CodeMirror already normalizes line endings to `\n` by default.

### CODE REVIEW

File: packages/codemirror/src/editor.ts [hunk 1]
```diff
@@ -716,6 +716,13 @@ class CodeMirrorEditor implements CodeEditor.IEditor {
        doc.replaceRange('', from, to);
        break;
      case 'set':
+       // Detect the line endings in the text and update
```
Comment (sccolbert) [inline packages/codemirror/src/editor.ts]: Should this be set just once when the editor is constructed?

Comment (blink1073) [reply packages/codemirror/src/editor.ts]: The document text is loaded asynchronously

Comment (sccolbert) [reply packages/codemirror/src/editor.ts]: Do we not even check at construction whether the string object is empty?

Comment (sccolbert) [reply packages/codemirror/src/editor.ts]: here: https://github.com/blink1073/jupyterlab/blob/259ff7aaf25261526d3252a2ab31d5cb2b397deb/packages/codemirror/src/editor.ts#L112

Comment (blink1073) [reply packages/codemirror/src/editor.ts]: True, we need to check in the constructor as well

File: packages/docregistry/src/context.ts [hunk 2]
```diff
@@ -280,13 +280,22 @@ class Context<T extends DocumentRegistry.IModel> implements DocumentRegistry.ICo
...
       if (contents.format === 'json') {
         model.fromJSON(contents.content);
       } else {
-        model.fromString(contents.content);
+        let content = contents.content;
+        // Convert line endings if necessary, marking the file
+        // as dirty.
+        if (content.indexOf('\r') !== -1) {
+          dirty = true;
+          content = content.replace(/\r\n/g, '\n');
```
Comment (sccolbert) [inline packages/docregistry/src/context.ts]: you can do this with a single regex:
`content.replace(/\r\n|\r/g, '\n')`

**Comment by ian-r-rose** [comment]
I also saw the issue with a document today (using Ubuntu), and can confirm that this fixes the problem.
Thanks to @blink1073 for chasing this down!

**Pull request merged**